Author: Eric Long <i@hack3r.moe>
Author: Andreas Beckmann <anbe@debian.org>
Description: fix building for riscv64
Origin: other, https://bugs.debian.org/1056007


Added src/nvcore/Debug.h section.

Signed-off-by: Marcin Juszkiewicz <mjuszkiewicz@redhat.com>


Index: nvidia-texture-tools-2.1.2/extern/poshlib/posh.h
===================================================================
--- nvidia-texture-tools-2.1.2.orig/extern/poshlib/posh.h
+++ nvidia-texture-tools-2.1.2/extern/poshlib/posh.h
@@ -565,6 +565,17 @@ LCC predefines the following:
 #  define POSH_CPU_STRING "S390"
 #endif
 
+#if defined __riscv
+#  define POSH_CPU_RISCV 1
+#  if __riscv_xlen == 64
+#    define POSH_CPU_RISCV64 1
+#    define POSH_CPU_STRING "riscv64"
+#  elif __riscv_xlen == 32
+#    define POSH_CPU_RISCV32 1
+#    define POSH_CPU_STRING "riscv32"
+#  endif
+#endif
+
 #if !defined POSH_CPU_STRING
 #  error POSH cannot determine target CPU
 #  define POSH_CPU_STRING "Unknown" /* this is here for Doxygen's benefit */
Index: nvidia-texture-tools-2.1.2/src/nvcore/Debug.cpp
===================================================================
--- nvidia-texture-tools-2.1.2.orig/src/nvcore/Debug.cpp
+++ nvidia-texture-tools-2.1.2/src/nvcore/Debug.cpp
@@ -668,6 +668,9 @@ namespace
 #    elif NV_CPU_AARCH64
         ucontext_t * ucp = (ucontext_t *)secret;
         return (void *) ucp->uc_mcontext.pc;
+#    elif NV_CPU_RISCV64
+        ucontext_t * ucp = (ucontext_t *)secret;
+        return (void *) ucp->uc_mcontext.__gregs[REG_PC];
 #   elif NV_CPU_E2K /* MCST Elbrus 2000 */
         // e2k has 3 stacks - Procedure Stack (PS), Procedure Chain Stack (PCS) and User Stack (US)
         // CR0 and CR1 (Chain Register) are the 128-bit registers of the Procedure Chain Stack (PCS)
Index: nvidia-texture-tools-2.1.2/src/nvcore/nvcore.h
===================================================================
--- nvidia-texture-tools-2.1.2.orig/src/nvcore/nvcore.h
+++ nvidia-texture-tools-2.1.2/src/nvcore/nvcore.h
@@ -113,6 +113,8 @@
 #   define NV_CPU_ARM 1
 #elif defined POSH_CPU_AARCH64
 #   define NV_CPU_AARCH64 1
+#elif defined POSH_CPU_RISCV64
+#   define NV_CPU_RISCV64 1
 #elif defined POSH_CPU_E2K
 #   define NV_CPU_E2K 1
 #elif defined POSH_CPU_MIPS
Index: nvidia-texture-tools-2.1.2/src/nvthread/Atomic.h
===================================================================
--- nvidia-texture-tools-2.1.2.orig/src/nvthread/Atomic.h
+++ nvidia-texture-tools-2.1.2/src/nvthread/Atomic.h
@@ -88,6 +88,16 @@ namespace nv {
 		uint32 ret = *ptr; // replace with ldrex?
         nvCompilerReadWriteBarrier();
         return ret;
+#elif POSH_CPU_RISCV64
+        // need more specific cpu type for riscv64?
+        // also utilizes a full barrier
+        // currently treating load like x86 - this could be wrong
+
+        // this is the easiest but slowest way to do this
+        nvCompilerReadWriteBarrier();
+        uint32 ret = *ptr; // replace with ldrex?
+        nvCompilerReadWriteBarrier();
+        return ret;
 #elif POSH_CPU_E2K
         // need more specific cpu type for e2k?
         // also utilizes a full barrier
@@ -127,6 +137,11 @@ namespace nv {
         nvCompilerReadWriteBarrier();
 		*ptr = value; //strex?
 		nvCompilerReadWriteBarrier();
+#elif POSH_CPU_RISCV64
+        // this is the easiest but slowest way to do this
+        nvCompilerReadWriteBarrier();
+	*ptr = value; //strex?
+	nvCompilerReadWriteBarrier();
 #elif POSH_CPU_E2K
         // this is the easiest but slowest way to do this
         nvCompilerReadWriteBarrier();
Index: nvidia-texture-tools-2.1.2/src/nvcore/Debug.h
===================================================================
--- nvidia-texture-tools-2.1.2.orig/src/nvcore/Debug.h
+++ nvidia-texture-tools-2.1.2/src/nvcore/Debug.h
@@ -166,7 +166,7 @@ NVCORE_API void NV_CDECL nvDebugPrint( c
 namespace nv
 {
     inline bool isValidPtr(const void * ptr) {
-    #if NV_CPU_X86_64 || POSH_CPU_PPC64 || NV_CPU_AARCH64 || NV_CPU_S390
+    #if NV_CPU_X86_64 || POSH_CPU_PPC64 || NV_CPU_AARCH64 || NV_CPU_S390 || NV_CPU_RISCV64
         if (ptr == NULL) return true;
         if (reinterpret_cast<uint64>(ptr) < 0x10000ULL) return false;
         if (reinterpret_cast<uint64>(ptr) >= 0x000007FFFFFEFFFFULL) return false;
